lib/realtime/syn_shards.ex

```diff
@@ -0,0 +1,59 @@
+defmodule Realtime.SynShards do
+  @moduledoc """
+  Implements sharding for Syn
+  """
+  @shards_num Application.compile_env(:realtime, :syn_shards)
+
+  @spec add_node_to_scopes(atom()) :: :ok
+  def add_node_to_scopes(scope) do
+    scope
+    |> scope_shards()
+    |> :syn.add_node_to_scopes()
+  end
+
+  @spec group_names(atom()) :: [term()]
+  def group_names(scope, node \\ false) do
+    scope
+    |> scope_shards()
+    |> Enum.reduce([], fn shard, acc ->
+      group_names =
+        if node,
+          do: :syn.group_names(shard, node),
+          else: :syn.group_names(shard)
+
+      group_names ++ acc
+    end)
+    |> Enum.uniq()
+  end
+
+  @spec join(atom(), term(), pid()) :: :ok | {:error, term()}
+  def join(scope, name, pid) do
+    scope
```

Comment:
Shard by pid instead? Otherwise a tenant will be pinned to one syn process.

Reply:
Damn, such a stupid mistake. Of course, sharing by name doesn’t make sense in our case 🤦‍♂️🤦‍♂️🤦‍♂️